### PR TITLE
Add types & boilerplate for the Downtime detector module

### PR DESCRIPTION
**Pull request by ValarDragon**
Sub-component of #3603 

## What is the purpose of the change

Adds types for the thin module intended for downtime detection

## Brief Changelog

- Add downtime detection module types

## Testing and Verifying

No tests added

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? somewhat
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? yes
  - How is the feature or change documented? In its spec

### CODE REVIEW

File: proto/osmosis/downtime-detector/v1beta1/genesis.proto [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import "cosmos_proto/cosmos.proto";
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/osmosis-labs/osmosis/v13/x/downtime-detector/types";
+
+// Params holds parameters for the downtime-detector module
+message Params {}
+
+message GenesisDowntimeEntry {
```
Comment (p0mvn) [inline proto/osmosis/downtime-detector/v1beta1/genesis.proto:15]: nit: consider calling it just "DowntimeEntry". Once the module launches, the "Genesis" prefix might feel confusing, especially if we want to make this be configurable.

**Review by p0mvn — APPROVED** [review]
LGTM

File: x/downtime-detector/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+In the downtime detector module, we store state entries for:
+
+* Last blocks timestamp
+* For each period, last time there was downtime
+
+Then in every begin block:
+
+* Store last blocks timestamp
+* if time since last block timestamp >= 30 seconds, iterate through all $DOWNTIME_PERIODS and add a state entry for the current block time
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/downtime-detector/README.md:28]: ah yeah, right! Should amend this, would iterate over all applicable durations that the downtime is longer than

File: proto/osmosis/downtime-detector/v1beta1/genesis.proto (hunk 1 above)
Comment (ValarDragon) [reply proto/osmosis/downtime-detector/v1beta1/genesis.proto:15]: But its not the live state entry format, hence the genesis prefix

File: proto/osmosis/downtime-detector/v1beta1/genesis.proto [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    (gogoproto.moretags) = "yaml:\"downtime_duration\""
+  ];
+  google.protobuf.Timestamp last_downtime = 2 [
+    (gogoproto.nullable) = false,
+    (gogoproto.stdtime) = true,
+    (gogoproto.moretags) = "yaml:\"last_downtime\""
+  ];
+}
+
+// GenesisState defines the twap module's genesis state.
```
Comment (czarcas7ic) [inline proto/osmosis/downtime-detector/v1beta1/genesis.proto:28]:
```suggestion
// GenesisState defines the downtime-detector module's genesis state.
```

File: proto/osmosis/downtime-detector/v1beta1/genesis.proto [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+message GenesisState {
+  repeated GenesisDowntimeEntry downtimes = 1 [ (gogoproto.nullable) = false ];
+
+  google.protobuf.Timestamp last_block_time = 2 [
+    (gogoproto.nullable) = false,
+    (gogoproto.stdtime) = true,
+    (gogoproto.moretags) = "yaml:\"last_block_time\""
+  ];
+
+  // params is the container of twap parameters.
```
Comment (czarcas7ic) [inline proto/osmosis/downtime-detector/v1beta1/genesis.proto:38]:
```suggestion
  // params is the container of downtime-detector parameters.
```

File: x/downtime-detector/downtimedetector_module/module.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+func (AppModuleBasic) Name() string { return types.ModuleName }
+
+func (AppModuleBasic) RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
+}
+
+func (AppModuleBasic) DefaultGenesis(cdc codec.JSONCodec) json.RawMessage {
+	return cdc.MustMarshalJSON(types.DefaultGenesis())
+}
+
+// ValidateGenesis performs genesis state validation for the gamm module.
```
Comment (czarcas7ic) [inline x/downtime-detector/downtimedetector_module/module.go:41]:
```suggestion
// ValidateGenesis performs genesis state validation for the downtime-detector module.
```

File: x/downtime-detector/downtimedetector_module/module.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+func (b AppModuleBasic) GetTxCmd() *cobra.Command {
+	return nil
+	// return cli.NewTxCmd()
+}
+
+func (b AppModuleBasic) GetQueryCmd() *cobra.Command {
+	return twapcli.GetQueryCmd()
+}
+
+// RegisterInterfaces registers interfaces and implementations of the gamm module.
```
Comment (czarcas7ic) [inline x/downtime-detector/downtimedetector_module/module.go:68]:
```suggestion
// RegisterInterfaces registers interfaces and implementations of the downtime-detector module.
```

File: x/downtime-detector/downtimedetector_module/module.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+func (AppModule) QuerierRoute() string { return types.RouterKey }
+
+func (am AppModule) LegacyQuerierHandler(legacyQuerierCdc *codec.LegacyAmino) sdk.Querier {
+	return func(sdk.Context, []string, abci.RequestQuery) ([]byte, error) {
+		return nil, fmt.Errorf("legacy querier not supported for the x/%s module", types.ModuleName)
+	}
+}
+
+// InitGenesis performs genesis initialization for the twap module.
```
Comment (czarcas7ic) [inline x/downtime-detector/downtimedetector_module/module.go:104]:
```suggestion
// InitGenesis performs genesis initialization for the downtime-detector module.
```

File: x/downtime-detector/downtimedetector_module/module.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+func (am AppModule) InitGenesis(ctx sdk.Context, cdc codec.JSONCodec, gs json.RawMessage) []abci.ValidatorUpdate {
+	var genesisState types.GenesisState
+
+	cdc.MustUnmarshalJSON(gs, &genesisState)
+
+	am.k.InitGenesis(ctx, &genesisState)
+	return []abci.ValidatorUpdate{}
+}
+
+// ExportGenesis returns the exported genesis state as raw bytes for the twap.
```
Comment (czarcas7ic) [inline x/downtime-detector/downtimedetector_module/module.go:115]:
```suggestion
// ExportGenesis returns the exported genesis state as raw bytes for the downtime-detector
```

File: x/downtime-detector/keeper.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+	// k.SetParams(ctx, genState.Params)
+}
+
+// ExportGenesis returns the twap module's exported genesis.
+func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
+	return &types.GenesisState{
+		// Params: k.GetParams(ctx),
+	}
+}
```
Comment (czarcas7ic) [inline x/downtime-detector/keeper.go:52]:
```suggestion
// GetParams returns the total set of downtime-detector parameters.
// func (k Keeper) GetParams(ctx sdk.Context) (params types.Params) {
// 	k.paramSpace.GetParamSet(ctx, &params)
// 	return params
// }

// // SetParams sets the total set of downtime-detector parameters.
// func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
// 	k.paramSpace.SetParamSet(ctx, &params)
// }

// InitGenesis initializes the downtime-detector module's state from a provided genesis
// state.
func (k Keeper) InitGenesis(ctx sdk.Context, genState *types.GenesisState) {
	if err := genState.Validate(); err != nil {
		panic(err)
	}

	// k.SetParams(ctx, genState.Params)
}

// ExportGenesis returns the downtime-detector module's exported genesis.
func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
	return &types.GenesisState{
		// Params: k.GetParams(ctx),
	}
}
```

**Review by czarcas7ic — APPROVED** [review]

File: x/downtime-detector/downtimedetector_module/module.go (hunk 5 above)
Comment (ValarDragon) [reply x/downtime-detector/downtimedetector_module/module.go:41]: im just going to delete these comments instead, they're generally boilerplate

**Pull request merged**